### PR TITLE
Add Table of Content Support

### LINKED ISSUE #26: [Feature request] Table of content support
For ebooks with properly embedded TOC, any of these features would be nice:
- displaying a number of characters left in the current chapter
- a menu or sidebar with a TOC allowing for quick navigation at any time, preferably displaying the character counts at which the chapters begin, i.e. something like this: 
```
プロローグ	178
第一章　母親と少年	2607
第二章　告白と困惑	13768
etc.
```

### PR DESCRIPTION
**Pull request by Renji-XD**
closes #26 

- adds support for book internal toc in pagination mode and fix an ux issue for continuation mode in regards with configured margins
- adds keybinds for next/previous chapter on desktop
- as discussed before no fallback / retrospective build, only new / reimported books will store the required data
- the data is based on the epub navigation document (or ncx for v2) - if it contains invalid html/xml or is simply empty there will also be no sections stored (even if there is maybe an toc in the html files). Also Chapter Labels may differ to the html one
- as discussed before progression is based on element position and not on current explored characters
- similar to explored character count the progress is based on the right/top area of the element and will therefore typically not reach 100% of a book etc.

Kind of known limitation is it does not really work well for books using subchapters inside a section as the opf will only list the the parent item once in the spine and therefore stuff gets merged instead of being an separate entry

https://renji-xd.github.io/ebook-reader

### CODE REVIEW

File: apps/web/src/lib/components/book-reader/book-reader-continuous/book-reader-continuous.svelte [hunk 1]
```diff
@@ -111,6 +124,14 @@
 
   const destroy$ = new Subject<void>();
 
+  const sectionToElement: Map<string, HTMLElement> = new Map();
+
+  const sectionData: Map<string, SectionWithProgress> = new Map();
```
Comment (ChaseOxide) [inline apps/web/src/lib/components/book-reader/book-reader-continuous/book-reader-continuous.svelte:129]: For future reference you can actually do `const sectionData = new Map<string, SectionWithProgress>();`

File: apps/web/src/lib/components/book-reader/book-reader-paginated/page-manager-paginated.ts [hunk 2]
```diff
@@ -174,4 +196,33 @@ export class PageManagerPaginated implements PageManager {
       return subscription;
     });
   }
+
+  private updateSectionData(ref: string, progress: number, emit = true) {
+    if (ref && this.sectionData.has(ref)) {
```
Comment (ChaseOxide) [inline apps/web/src/lib/components/book-reader/book-reader-paginated/page-manager-paginated.ts:201]: For such code I usually prefer early return, like
`if (!ref || !this.sectionData.has(ref)) return`

**Review by ChaseOxide — CHANGES_REQUESTED** [review]
- ToC button should be in the top header's most left item
- I think the ToC popup's background colors are reversed (expect darker background on the book content, popup itself remain "natural")

What I've found so far

**Comment by Renji-XD** [comment]
> * ToC button should be in the top header's most left item
> * I think the ToC popup's background colors are reversed (expect darker background on the book content, popup itself remain "natural")

 - Not sure how much it is used, but i thought one click less is more convenient for a on page feature (therefore also no keybind) and it may also looks better than having an overlapping animation from the header. up to you
 - I am not sure what you mean with "natural" - i though it make sense to have a more darker background for a darker theme and vice versa. maybe just drop me the colors you prefer for dark / light mode and i can replace them

**Comment by ChaseOxide** [comment]
> Not sure how much it is used, but i thought one click less is more convenient for a on page feature (therefore also no keybind) and it may also looks better than having an overlapping animation from the header.
It took me few minutes to find the icon, just think the button is in an unnatural position, since from my experience these buttons are usually placed at the top.

> Also the header is kind of relatively full on mobile already - moving the icon there will not leave much place for much more in the future
If space is needed we'll have to make a dropdown, I still prefer more clicks over putting it at the bottom

For the pop out background, the sidebar background should be the same as the book's background color (if you're trying to stick with the reader's theme). Backdrop background should be the same as the dialog's (bg-black/[.32]), with the elevation of 16 (mdc-elevation--z16).
(Kind of hinting it should use dialogManager, but I think it's fine if it's separate)

**Pull request merged**